File: python3.6/BallGame.py
from threading import Thread
import time, random, math
from Ball import *


class BallGame(Thread):

    def __init__(self, ballsContainer, scorePanel):
        Thread.__init__(self)

        self.daemon                 = True
        self.running                = False
        self.balls                  = []
        self.ballsContainer         = ballsContainer
        self.ballsToAdd             = 0
        self.ballsToDelete          = 0
        self.reset                  = False
        self.scorePanel             = scorePanel
        self.score                  = 0

    def startGame(self):
        self.running = True;

    def stopGame(self):
        self.running = False;

    def addOneBall(self):
        if(len(self.balls) < 21):
            self.ballsToAdd += 1

    def deleteOneBall(self):
        self.ballsToDelete += 1

    def run(self):
        while True:
            # Suppression des balles
            while(self.ballsToDelete > 0):
                # S'il y a des balles à supprimer
                if(len(self.balls) > 0):
                    self.ballsContainer.delete(self.balls[-1].graphics)
                    self.balls.remove(self.balls[-1])
                    self.ballsToDelete -= 1
                else:
                    self.ballsToDelete = 0

            # Ajout des balles
            while(self.ballsToAdd > 0):

                self.ballsToAdd -= 1;
                r       = random.randrange(10,50)

                # On s'assure que ça n'apparaisse jamais sur un bord
                x       = random.randrange(0 + r, self.ballsContainer.winfo_width() - r)
                y       = random.randrange(0 + r, self.ballsContainer.winfo_height() - r)

                c_r     = hex(random.randrange(0,255))[2:]
                if(len(c_r) < 2):
                    c_r = "0" + c_r

                c_v     = hex(random.randrange(0,255))[2:]
                if(len(c_v) < 2):
                    c_v = "0" + c_v

                c_b     = hex(random.randrange(0,255))[2:]
                if(len(c_b) < 2):
                    c_b = "0" + c_b


                color = "#" + str(c_r) + "" + str(c_v) + "" + str(c_b)

                # On crée la balle
                b = Ball(x,y,r)

                b.setColor(color)

                b.setSpeed(random.randrange(2,10))

                b.setDirection(random.randrange(0,360))

                bGraphics = self.ballsContainer.create_oval(b.x-b.r, b.y-b.r, b.x+b.r, b.y+b.r, fill=b.color)

                b.setGraphics(bGraphics)

                self.balls.append(b)

            if(self.running):
                self.drawBalls()
                self.moveBalls()
                self.ballsCollision()
            time.sleep(0.015)

    def ballsCollision(self):
        toDelete = []
        for ballFirst in self.balls:
            for ballSecond in self.balls:

                if(ballFirst != ballSecond):
                    dx = ballFirst.x - ballSecond.x
                    dy = ballFirst.y - ballSecond.y

                    dist = math.sqrt(dx * dx + dy * dy)

                    if(dist < ballFirst.r + ballSecond.r):
                        if(ballFirst not in toDelete):
                            toDelete.append(ballFirst)
                        if(ballSecond not in toDelete):
                            toDelete.append(ballSecond)

        tempScore = 0
        for ball in toDelete:
            self.ballsContainer.delete(ball.graphics)
            self.balls.remove(ball)
            tempScore += 1

        if(tempScore > 0):
            self.score += int(tempScore/2)
            self.scorePanel.config(text="Score : " + str(self.score))



    def resetGame(self):
        self.score = 0
        self.scorePanel.config(text="Score : 0")
        toDelete = []
        for ball in self.balls:
            toDelete.append(ball)

        # Ici on utilise une autre liste
        # Demande de la mémoire et du cpu, mais négligeable au vu du maximum de balles
        for ball in toDelete:
            self.ballsContainer.delete(ball.graphics)
            self.balls.remove(ball)


    # Deplace l'ensemble des balles
    def moveBalls(self):
        for ball in self.balls:
            if(ball.x - ball.r - 1) <= 0:
                ball.invertDx()
            if(ball.x + ball.r + 1) >= self.ballsContainer.winfo_width():
                ball.invertDx()
            if(ball.y - ball.r - 1) <= 0:
                ball.invertDy()
            if(ball.y + ball.r + 1) >= self.ballsContainer.winfo_height():
                ball.invertDy()

            ball.move()

    # Affiche l'ensemble des balles à l'écran
    def drawBalls(self):
        for ball in self.balls:
            self.ballsContainer.move(ball.graphics, ball.d_x, ball.d_y)

File: Java/BallGame.java
package balls;

import java.awt.Color;
import java.awt.Graphics;
import java.util.Random;
import java.util.Vector;

import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;

public class BallGame extends Thread {
	boolean running = false;
	BallsContainer ballsArea;
	JButton plusButton,minusButton,startStopButton;
	Random rModule = new Random();
	int addOneBall = 0;
	int deleteOneBall = 0;
	JLabel scorePanel;
	boolean reset = false;
	
	public BallGame(BallsContainer ballsArea) {
		this.ballsArea 			= ballsArea;
		this.setDaemon(true);
	}
	
	public void setScorePanel(JLabel score) {
		scorePanel = score;
	}
	
	public void startGame() {
		running = true;
	}
	
	public void stopGame() {
		running = false;
	}
	
	public void run() {
		try {
			while(true) {
				
				while(deleteOneBall > 0) {
					ballsArea.deleteBall();
					deleteOneBall--;
					ballsArea.repaint();
				}
				
				while(addOneBall > 0) {
					addOneBall--;
					int x = rModule.nextInt(ballsArea.getWidth());
					int y = rModule.nextInt(ballsArea.getHeight());
					// Radius 10 min 60 max
					int r = rModule.nextInt(50) + 40;
					if(x+r > ballsArea.getWidth()) x -= r;
					if(y+r > ballsArea.getHeight()) y -= r;
					Ball nBall = new Ball(x, y, r);
					nBall.setColor(new Color(rModule.nextInt(255), rModule.nextInt(255), rModule.nextInt(255)));
					nBall.setSpeed(rModule.nextInt(14) + 2);
					nBall.setDirection(rModule.nextInt(360));
					nBall.updateDeplacement();
					ballsArea.addBall(nBall);
					ballsArea.repaint();
				}
				
				if(running) {
					ballsArea.ballsCollision();
					ballsArea.moveBalls();
					scorePanel.setText("Score : " + ballsArea.score);
					ballsArea.repaint();
				}
				
				if(reset) {
					reset = false;
					ballsArea.reset();
				}
				
				sleep((long) 16.67);
			}
		} catch(InterruptedException e) {}
	}
	
	public void addOneBall() {
		if(!ballsArea.ballsFull()) {
			addOneBall += 1;
		}
	}
	
	public void removeOneBall() {
		if(ballsArea.atLeastOneBall()) {
			deleteOneBall++;
		}	
	}
	
	public void resetGame() {
		reset = true;
	}

}
